src/script/miniscript.h

```diff
@@ -0,0 +1,998 @@
+
+    to_parse.emplace_back(ParseContext::WRAPPED_EXPR, -1, -1);
+
+    while (!to_parse.empty()) {
+        // Get the current context we are decoding within
+        auto [cur_context, n, k] = to_parse.back();
+        to_parse.pop_back();
+
+        switch (cur_context) {
+        case ParseContext::WRAPPED_EXPR: {
+            int colon_index = -1;
+            for (int i = 1; i < (int)in.size(); ++i) {
+                if (in[i] == ':') {
+                    colon_index = i;
+                    break;
+                }
+                if (in[i] < 'a' || in[i] > 'z') break;
+            }
+            // If there is no colon, this loop won't execute
+            for (int j = 0; j < colon_index; ++j) {
+                if (in[j] == 'a') {
+                    to_parse.emplace_back(ParseContext::ALT, -1, -1);
+                } else if (in[j] == 's') {
+                    to_parse.emplace_back(ParseContext::SWAP, -1, -1);
+                } else if (in[j] == 'c') {
+                    to_parse.emplace_back(ParseContext::CHECK, -1, -1);
+                } else if (in[j] == 'd') {
+                    to_parse.emplace_back(ParseContext::DUP_IF, -1, -1);
+                } else if (in[j] == 'j') {
+                    to_parse.emplace_back(ParseContext::NON_ZERO, -1, -1);
+                } else if (in[j] == 'n') {
+                    to_parse.emplace_back(ParseContext::ZERO_NOTEQUAL, -1, -1);
+                } else if (in[j] == 'v') {
+                    to_parse.emplace_back(ParseContext::VERIFY, -1, -1);
+                } else if (in[j] == 'u') {
+                    to_parse.emplace_back(ParseContext::WRAP_U, -1, -1);
+                } else if (in[j] == 't') {
+                    to_parse.emplace_back(ParseContext::WRAP_T, -1, -1);
+                } else if (in[j] == 'l') {
+                    // The l: wrapper is equivalent to or_i(0,X)
+                    constructed.push_back(MakeNodeRef<Key>(NodeType::JUST_0));
+                    to_parse.emplace_back(ParseContext::OR_I, -1, -1);
+                } else {
+                    return {};
+                }
+            }
+            to_parse.emplace_back(ParseContext::EXPR, -1, -1);
+            in = in.subspan(colon_index + 1);
+            break;
+        }
+        case ParseContext::EXPR: {
+            if (Const("0", in)) {
+                constructed.push_back(MakeNodeRef<Key>(NodeType::JUST_0));
+            } else if (Const("1", in)) {
+                constructed.push_back(MakeNodeRef<Key>(NodeType::JUST_1));
+            } else if (Const("pk(", in)) {
+                Key key;
+                int key_size = FindNextChar(in, ')');
+                if (key_size < 1) return {};
+                if (!ctx.FromString(in.begin(), in.begin() + key_size, key)) return {};
+                constructed.push_back(MakeNodeRef<Key>(NodeType::WRAP_C, Vector(MakeNodeRef<Key>(NodeType::PK_K, Vector(std::move(key))))));
+                in = in.subspan(key_size + 1);
```

Comment:
In 3a2c00d9bc3c7a34fb8727d76e9c3a691395b67c "Miniscript: type system, script creation, text notation, tests"

Can the key parsing stuff be refactored into its own function instead of copy and pasted for the next 3 expressions?